text/0004-recoverable-transaction-history.md

```diff
@@ -0,0 +1,351 @@
+
+(See in code: https://github.com/mobilecoinfoundation/mobilecoin/pull/814/files#diff-681e5a738ac3ba20d8bcb29bcd66424b189075a4a2c1a27b9429b9da78aa7946R14)
+
+Clients should interpret this memo as indicating who sent them this payment,
+once it has been validated. If the address hash is unknown, or the validation fails,
+then they should not associate this payment with a sender.
+
+## 0x0101 Authenticated Sender With Payment Request Id Memo
+
+The 44-byte memo data is laid out as follows:
+
+| Byte range | Item |
+| ---------- | ---- |
+| 0 - 16     | Sender's Address Hash |
+| 16 - 24    | Big-endian bytes of 8-byte payment request id number |
+| 24 - 28    | Unused bytes          |
+| 28 - 44    | HMAC                  |
+
+The HMAC is computed in the same way as for 0x0100 Authenticated Sender Memo.
+
+Clients may interpret this memo in the same way as the 0x0100 Authenticated Sender Memo,
+and additionally conclude that the sender wished to associate this payment to a specific
+payment request.
+
+There is an additional way that the client can handle this memo which we wish to call out:
+
+In some payment flows, Bob would like to request payment from Alice, but Bob does not
+yet have Alice's public address. We propose that as long as Bob chooses a (secure) uniformly random
+64-bit payment request id and includes this in the payment request that he sends to Alice,
+Bob's client may conclude that a payment came from Alice based on matching the value in
+the payment request id field, even if he does not check the HMAC (he cannot do that
+if he doesn't yet have Alice's public address). Later, if Bob's device is lost, he may
+lose track of these payment request id numbers, but can still associate this payment to Alice
+by validating the memo along the normal pathway if at that later time she has entered his contacts.
+
+In such a case, it is acceptable to interpret the memo and access the payment request id
+number, even if we haven't validated the memo by confirming the HMAC.
+
+## 0x0200 Destination Memo
+
+The 44-byte memo data is laid out as follows:
+
+| Byte range | Item |
+| ---------- | ---- |
+| 0 - 16     | Recipient's Address Hash |
+| 16 - 17    | The number of recipients, as an unsigned 8-bit number |
+| 17 - 24    | Big-endian bytes of fee amount, as an unsinged 56-bit number |
+| 24 - 32    | Big-endian bytes of the total outlay amount, as an unsigned 64-bit number |
+| 32 - 44    | Unused bytes                  |
+
+Here, the "total outlay" means the total amount that this transaction is deducting
+from the Sender's balance. It is the sum of the value of all non-change outputs,
+and the fee.
+
+The destination memo imposes some limits:
+* If the number of recipients exceeds 255 then the memo cannot be used. Since there are only 16 outputs
+  per transaction at most at time of writing, this is a safe assumption.
+* If there is more than one recipient, only one address hash can be recorded.
+  The client should choose one of them.
```

Comment:
Any reason not to just use a single recipient here, and just say "destination memos can only be used with single-recipient txes", since it kind of falls apart if you don't?

Would that leave space for a MAC that would remove the need to keep the change address secret?

Reply:
It would save us exactly one byte, and we are cannibalizing a byte from the fee to get that byte already

We have 12 unused bytes right now. If we use 16 bytes on a MAC here then we won't have any spare bytes to do something like, record payment request ID or refund ID or anything else which people may want later, so it seems more future proof if we use the subaddress validation idea that Kyle came up with